### PR TITLE
Support to upgrade local chart app

### PR DESCRIPTION
**Pull request by thxCode**
**Problem:**
- Cannot upgrade local chart app
- `app show-app`, `app ls` cannot indicate the chart information

**Solution:**
- Check local template path exists or not when upgrading
- Fetch chart information from catalog `ExternalID` or local chart template `Files`

**Issue:**
rancher/rancher#15711

**Related PR:**
- https://github.com/rancher/types/pull/664
- https://github.com/rancher/rancher/pull/17278

### CODE REVIEW

**Comment by thxCode** [comment]
After enriching app commands:

```bash
$ rancher app ls
ID                    NAME          STATE     CATALOG        TEMPLATE   VERSION
p-ggq9l:example-app   example-app   active    local folder   example2   0.2.0
p-ggq9l:test-redis    test-redis    active    library        redis      3.0.2

$ rancher app sa example-app
CURRENT   REVISION            CATALOG        TEMPLATE   VERSION   CREATED
*         apprevision-j5v59   local folder   example2   0.2.0     04 Jan 2019 02:03:59 UTC
          apprevision-4h9ll   local folder   example    0.1.0     04 Jan 2019 02:02:37 UTC

$ rancher app sa test-redis
CURRENT   REVISION            CATALOG   TEMPLATE   VERSION   CREATED
*         apprevision-rrhzb   library   redis      3.0.2     02 Jan 2019 05:47:43 UTC

CURRENT   VERSION
*         3.0.2

```

File: cmd/app.go [hunk 1]
```diff
@@ -323,30 +325,49 @@ func appUpgrade(ctx *cli.Context) error {
...
-		return fmt.Errorf("version %s not valid for app", ctx.Args().Get(1))
-	}
+		q := u.Query()
+		q.Set("version", appVersion)
 
-	au := &projectClient.AppUpgradeConfig{
-		Answers:    answers,
-		ExternalID: template.Data[0].ExternalID,
+		filter := defaultListOpts(ctx)
+		filter.Filters["externalId"] = u.String()
```
Comment (gitlawr) [inline cmd/app.go]: The version in `u.String()` is the original one, not the version about to upgrade.

File: cmd/app.go [hunk 2]
```diff
@@ -323,30 +325,49 @@ func appUpgrade(ctx *cli.Context) error {
...
+		template, err := c.ManagementClient.TemplateVersion.List(filter)
+		if err != nil {
+			return err
+		}
+
+		if len(template.Data) == 0 {
+			return fmt.Errorf("version %s not valid for app template %q", appVersion, q.Get("template"))
+		}
+
+		au.ExternalID = u.String()
```
Comment (gitlawr) [inline cmd/app.go]: ditto

File: cmd/app.go [hunk 3]
```diff
@@ -238,30 +260,90 @@ func appLs(ctx *cli.Context) error {
...
-		parsedExternal, err := parseExternalID(item.ExternalID)
+		appData := &AppData{
+			ID:  item.ID,
+			App: item,
+		}
+
+		appExternalID := item.ExternalID
+		var appTemplateFiles map[string]string
+		if appExternalID == "" {
+			appRevision, err := c.ProjectClient.AppRevision.ByID(strings.Replace(item.ID, item.Name, item.AppRevisionID, -1))
```
Comment (gitlawr) [inline cmd/app.go]: I know this is for generating a namespaced revisionId whereas we don't have a namespace prefix in the appRevisionID field. But it looks confusing. Extract the `strings.Replace` to a single line and add a comment for it, or fetch the namespace from `projectID` would be more descriptive.

File: cmd/app.go [hunk 4]
```diff
@@ -815,6 +917,23 @@ func parseExternalID(e string) (map[string]string, error) {
 	return parsed, nil
 }
 
+// parseChartInfo gives back is local folder deployment or not and a map with the keys catalog, template and version
+func parseChartInfo(c string) (bool, map[string]string, error) {
```
Comment (gitlawr) [inline cmd/app.go]: This function has no reference anywhere.

File: cmd/app.go (hunk 1 above)
Comment (thxCode) [reply cmd/app.go]: Fixed

File: cmd/app.go (hunk 2 above)
Comment (thxCode) [reply cmd/app.go]: Fixed

File: cmd/app.go (hunk 4 above)
Comment (thxCode) [reply cmd/app.go]: Removed

File: cmd/app.go (hunk 3 above)
Comment (thxCode) [reply cmd/app.go]: Adjusted

File: cmd/app.go [hunk 5]
```diff
@@ -89,7 +110,7 @@ type chartMetadata struct {
 	Icon        string   `json:"icon,omitempty" yaml:"icon,omitempty"`
 }
 
-type revSlice []revision
+type revSlice []*revision
```
Comment (gitlawr) [inline cmd/app.go]: Why do we need this change?

File: cmd/app.go [hunk 6]
```diff
@@ -238,30 +260,93 @@ func appLs(ctx *cli.Context) error {
...
-			Version: parsedExternal["version"],
-		})
+
+		writer.Write(appData)
 	}
 	return writer.Err()
 
 }
 
+func parseTemplateInfo(appExternalID string, appTemplateFiles map[string]string) (templateCatalog, templateName, templateVersion string, err error) {
```
Comment (dramich) [inline cmd/app.go]: Have this function return a map like `parseExternalID` and an error. Naked returns should be for short functions that are easy to follow.

File: cmd/app.go [hunk 7]
```diff
@@ -238,30 +260,93 @@ func appLs(ctx *cli.Context) error {
...
+
+			appRevision, err := c.ProjectClient.AppRevision.ByID(appRevisionID)
+			if err != nil {
+				return err
+			}
+
+			appTemplateFiles = appRevision.Status.Files
+		}
+
+		appData.Catalog, appData.Template, appData.Version, err = parseTemplateInfo(appExternalID, appTemplateFiles)
```
Comment (dramich) [inline cmd/app.go]: Use the map returned by `parseTemplateInfo` to populate `appData`

File: cmd/app.go [hunk 8]
```diff
@@ -769,14 +860,28 @@ func outputRevisions(ctx *cli.Context, c *cliclient.MasterClient) error {
...
 			return err
 		}
-		sorted = append(sorted, revision{Name: rev.Name, Created: parsedTime})
+
+		reversionData := revision{
+			Name:    rev.Name,
+			Created: parsedTime,
+		}
+
+		reversionData.Catalog, reversionData.Template, reversionData.Version, err = parseTemplateInfo(rev.Status.ExternalID, rev.Status.Files)
```
Comment (dramich) [inline cmd/app.go]: Use the map returned by `parseTemplateInfo` to populate `appData`

File: cmd/app.go [hunk 9]
```diff
@@ -637,6 +610,67 @@ func resolveTemplatePath(templateName string) bool {
 	return templateName == "." || strings.Contains(templateName, "\\\\") || strings.Contains(templateName, "/")
 }
 
+func walkTemplateDirectory(templatePath string) (chartName string, files map[string]string, err error) {
```
Comment (dramich) [inline cmd/app.go]: This function should not do naked returns, it is doing too much to easily grok what is happening with 2 nested anonymous functions.

File: cmd/app.go [hunk 10]
```diff
@@ -238,30 +260,93 @@ func appLs(ctx *cli.Context) error {
...
+			}
+
+			version := &chartVersion{}
+			unmarshalErr := yaml.Unmarshal(content, version)
+			if unmarshalErr != nil {
+				err = errors.Wrap(unmarshalErr, "can't parse Chart.yaml from app")
+				return
+			}
+
+			templateCatalog = "local folder"
```
Comment (dramich) [inline cmd/app.go]: nit: s/folder/directory - folder tends to mean more of a UI construct and a windows thing.

File: cmd/app.go (hunk 6 above)
Comment (thxCode) [reply cmd/app.go]: Refactored

File: cmd/app.go (hunk 9 above)
Comment (thxCode) [reply cmd/app.go]: Refactored

File: cmd/app.go [hunk 11]
```diff
@@ -649,6 +697,66 @@ func resolveTemplatePath(templateName string) bool {
...
+	templateAbsPath, parsedErr := filepath.Abs(templatePath)
+	if parsedErr != nil {
+		return "", nil, parsedErr
+	}
+	if _, statErr := os.Stat(templateAbsPath); statErr != nil {
+		return "", nil, statErr
+	}
+
+	chartName := ""
+	files := map[string]string{}
```
[the remark this answers is not in the document]
Comment (thxCode) [reply cmd/app.go]: Changed

File: cmd/app.go [hunk 12]
```diff
@@ -649,6 +697,66 @@ func resolveTemplatePath(templateName string) bool {
...
+func walkTemplateDirectory(templatePath string) (string, map[string]string, error) {
+	templateAbsPath, parsedErr := filepath.Abs(templatePath)
+	if parsedErr != nil {
+		return "", nil, parsedErr
+	}
+	if _, statErr := os.Stat(templateAbsPath); statErr != nil {
+		return "", nil, statErr
+	}
+
+	chartName := ""
```
[the remark this answers is not in the document]
Comment (thxCode) [reply cmd/app.go]: Got it

File: cmd/app.go [hunk 13]
```diff
@@ -238,30 +260,88 @@ func appLs(ctx *cli.Context) error {
...
+		{"TEMPLATE", "Template"},
 		{"VERSION", "Version"},
 	}, ctx)
 
 	defer writer.Close()
 
 	for _, item := range collection.Data {
-		parsedExternal, err := parseExternalID(item.ExternalID)
+		appExternalID := item.ExternalID
+		var appTemplateFiles map[string]string
```
[the remark this answers is not in the document]
Comment (thxCode) [reply cmd/app.go]: Refactored

**Comment by thxCode** [comment]
@dramich , Can you review again? I have addressed all your new comments.

**Comment by loganhz** [comment]
Thanks @dramich ! we will record these comments. we will fix it next time during the review from our side.

**Comment by dramich** [comment]
LGTM

**Pull request merged**